### DISCUSSION TITLE
Generate get & set methods automatically for protocol class

### DISCUSSION THREAD

**Discussion by Frank1016** [serverpod/serverpod#1457] [Ideas]
I'm working on a project using Serverpod as backend, and found that almost for all classes created in server protocol folder with a table, I have to manually create seperate getter and setter methods for each class in Flutter folder to easily manipulate the CRUD operations. When there are a lot of fields in the class, this becomes really a headache! Isn't it easy to generate associated getter & setter methods automatically for each class on the server side? Will this feature be added in Serverpod's newer version?

**Comment by vlidholt** [comment]
Hi,

I'm not exactly understanding which getter/setters you are talking about here. Are you talking about modifying a single property in the database? Otherwise, the properties can be modified right on the serializable objects, right?

**Reply by Frank1016** [reply]
Yes, I'm talking about modifying a single property for a class defined in server protocol folder. This happens quite often when you need to update something for a single row. I think adding setters method in the generated class file is the easiest way to do this kind of work, is there other easy way?

**Reply by Frank1016** [reply]
When there's only several properties for the class, it's easier to update values and no need for setters methods. However, when there are quite a lot of properties, it becomes a headache to update property values as you need to specify every single property name for setting values without setter method. If the setter method is generated automatically, all property names can be put in a map. In this way, when updating properties of this class, we can just pass a variable for property name in a loop, and no need to specify every single property name. For example, 

Without setter method:
`Future<void> update(ProtocolClass protocolClass) async {
  protocolClass
    ..propertyName1 = 'value1'
    ..propertyName2 = 'value2'
    ..propertyName3 = 'value3'
    ..propertyName4 = 'value4'
    ..propertyName5 = 'value5'
    ..propertyName6 = 'value6'
    ..propertyName7 = 'value7'
    ..propertyName8 = 'value8'
    ..propertyName9 = 'value9'
    ..propertyName10 = 'value10'
    ..propertyName11 = 'value11'
    ..propertyName12 = 'value12'
    ..propertyName13 = 'value13'
    ..propertyName14 = 'value14'
    ..propertyName15 = 'value15'
    ..propertyName16 = 'value16'
    ..propertyName17 = 'value17'
    ..propertyName18 = 'value18'
    ..propertyName19 = 'value19'
    ..propertyName20 = 'value20';

  await client.protocolClass.update(protocolClass);
}`

With setter method:
`Future<void> update(ProtocolClass protocolClass) async {
  for (var i = 1; i <= 20; i++) {
    protocolClass.set('propertyName$i', 'value$i');
  }
  await client.protocolClass.update(protocolClass);
}`

**Reply by Isakdl** [reply]
Hey @Frank1016!

This feature exists on main already! :) It is not in an official release yet though.

You can update a single column of a row like so:

```dart
await session.dbNext.updateRow<Model>(row, columns: [Model.t.column]);
```
Or even in an atomic batch update like so:

```dart
await session.dbNext.update<Model>(rows, columns: [Model.t.column]);
```

`copyWith` is also implemented and generated for all models, meaning you can selectively change a single value in any model.

```dart
var data = Model(...);
var copy = data.copyWith(fieldName: value);
```

Issue tangent to this: https://github.com/serverpod/serverpod/issues/1422

**Reply by Disregard-Therest** [reply]
Hi, what should I do to get copyWith enabled for my project? I don't see such functionality for my classes

**Reply by SandPod** [reply]
@Disregard-Therest  The `copyWith` functionality is included with Serverpod 1.2 that is released on Monday, countdown here: https://www.youtube.com/watch?v=AmHZP6T1NdQ

Once you have upgraded to Serverpod 1.2 you just need to regenerate your project and the models will be generated with the `copyWith` method.

The functionality is already available on the Serverpod `main`. If you can't wait until Monday you can also clone the branch and activate it locally, instructions here: https://docs.serverpod.dev/contribute#command-line-tools

**Reply by Disregard-Therest** [reply]
Great, thank you. Waiting for the release.
Looking forward to have explicit upgrade instructions.
